zipline/pipeline/domain.py

```diff
@@ -0,0 +1,182 @@
+
+from .sentinels import NotSpecified
+
+
+class IDomain(Interface):
+    """Domain interface.
+    """
+
+    @property
+    def name(self):
+        """User-facing name for this domain.
+        """
+
+    @property
+    def country_code(self):
+        """Country code for assets on this domain.
+        """
+
+    # TODO_SS: The original design for domains was to have them return a
+    # TradingCalendar, but we have a bunch of tests in test_blaze that use very
+    # short session indices that I don't know how to port to using a proper
+    # TradingCalendar.
+    #
+    # Is there a strong reason to prefer just exposing the calendar
+    # vs. exposing the sessions? If so, what do we do about the blaze tests?
+    def get_sessions(self):
+        """Get all trading sessions for the calendar of this domain.
+        """
+
+
+# Create a base class so that we can do type-based validation of user input.
+Domain = implements(IDomain)
+Domain.__name__ = 'Domain'
+
+
+# TODO: Better name for this?
+# TODO: Do we want/need memoization for this?
+class StandardDomain(Domain):
+    """TODO_SS
+    """
+    def __init__(self, name, country_code, calendar_name):
+        self._name = name
+        self._country_code = country_code
+        self._calendar_name = calendar_name
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def country_code(self):
+        return self._country_code
+
+    def get_sessions(self):
+        return self.calendar.all_sessions
+
+    @lazyval
+    def calendar(self):
+        return get_calendar(self._calendar_name)
+
+    def __str__(self):
+        return self._name
+
+    def __repr__(self):
+        return "StandardDomain(name={!r}, country={!r}, calendar={!r})".format(
+            self._name,
+            self.country_code,
+            self._calendar_name,
+        )
+
+
+# TODO: Is this the casing convention we want for domains?
+USEquities = StandardDomain('USEquities', CountryCode.UNITED_STATES, 'NYSE')
+CanadaEquities = StandardDomain('CanadaEquities', CountryCode.CANADA, 'TSX')
+# XXX: The actual country code for this is GB. Should we use that for the name
+# here?
+UKEquities = StandardDomain('UKEquities', CountryCode.UNITED_KINGDOM, 'LSE')
+
+
+def infer_domain(terms):
+    """
+    Infer the domain from a collection of terms.
+
+    The algorithm for inferring domains is as follows:
+
+    - If all input terms have a domain of NotSpecified, the result is also
+      NotSpecified.
+
+    - If there is exactly one non-NotSpecified domain in the input terms, the
+      result is that domain.
+
+    - Otherwise, an AmbiguousDomain error is raised.
+
+    Parameters
+    ----------
+    terms : iterable[zipline.pipeline.term.Term]
+
+    Returns
+    -------
+    inferred : Domain or NotSpecified
+
+    Raises
+    ------
+    AmbiguousDomain
+        Raised if more than one concrete domain is present in the input terms.
+    """
+    domains = {NotSpecified}
```

Comment:
It makes the logic below simpler. If we initialized this to an empty set, then we'd have to check the `len(domains) == 0` case to handle an empty input, and in the `len(domains) == 1` or `len(domains) == 2` cases, we'd have to have sub-branches depending on whether or not `NotSpecified` was in the inputs (e.g., it's an error to have two different non-NotSpecified domains in the input).

I think a reasonable alternative implementation would be:

```python
def infer_domain(terms):
    domains = {}
    for t in terms:
        domains.update(t.domain)

    try:
        domains.remove(NotSpecified)
    except KeyError:
        pass

    if len(domains) == 0:
        return NotSpecified
    elif len(domains) == 1:
        return domains.pop()
    else:
        raise AmbiguousDomain(...)
```

These feel like they have about the same amount of complexity in them. Maybe the latter is slightly better since it only requires one call to `remove`.